peer/peerwg/wireguard.go

```diff
@@ -0,0 +1,416 @@
+	defer cancel()
+	defer c.Close()
+
+	dialAddrStr := net.JoinHostPort("127.0.0.1", strconv.Itoa(int(port)))
+	var stdDialer net.Dialer
+	server, err := stdDialer.DialContext(ctx, "tcp", dialAddrStr)
+	if err != nil {
+		wn.logger.Debug(ctx, "dial local port", slog.F("port", port), slog.Error(err))
+		return
+	}
+	defer server.Close()
+
+	connClosed := make(chan error, 2)
+	go func() {
+		_, err := io.Copy(server, c)
+		connClosed <- err
+	}()
+	go func() {
+		_, err := io.Copy(c, server)
+		connClosed <- err
+	}()
+	err = <-connClosed
+	if err != nil {
+		wn.logger.Debug(ctx, "proxy connection closed with error", slog.Error(err))
+	}
+	wn.logger.Debug(ctx, "forwarded connection closed", slog.F("local_addr", dialAddrStr))
+}
+
+func (wn *WireguardNetwork) Close() error {
+	_ = wn.Netstack.Close()
+	wn.Engine.Close()
+
+	return nil
+}
+
+// AddPeer adds a peer to the network from a WireguardPeerMessage. After adding
+// a peer, they may connect to you.
+func (wn *WireguardNetwork) AddPeer(peer WireguardPeerMessage) error {
+	wn.mu.Lock()
+	defer wn.mu.Unlock()
+
+	// If the peer already exists in the network map, do nothing.
+	for _, p := range wn.netMap.Peers {
+		if p.Key == peer.Public {
+			wn.logger.Debug(context.Background(), "peer already in netmap", slog.F("peer", peer.Public.ShortString()))
+			return nil
+		}
+	}
+
+	// The Tailscale engine owns this slice, so we need to copy to make
+	// modifications.
+	peers := append(([]*tailcfg.Node)(nil), wn.netMap.Peers...)
+
+	id, stableID := nodeIDs(peer.Public)
+	peers = append(peers, &tailcfg.Node{
+		ID:         id,
+		StableID:   stableID,
+		Name:       peer.Public.String() + ".com",
+		Key:        peer.Public,
+		DiscoKey:   peer.Disco,
+		Addresses:  []netaddr.IPPrefix{netaddr.IPPrefixFrom(peer.IPv6, 128)},
+		AllowedIPs: []netaddr.IPPrefix{netaddr.IPPrefixFrom(peer.IPv6, 128)},
+		DERP:       DefaultDerpHome,
+		Endpoints:  []string{DefaultDerpHome},
+	})
+
+	wn.netMap.Peers = peers
+
+	cfg, err := nmcfg.WGCfg(wn.netMap, logf, netmap.AllowSingleHosts|netmap.AllowSubnetRoutes, tailcfg.StableNodeID("nBBoJZ5CNTRL"))
+	if err != nil {
+		return xerrors.Errorf("create wgcfg: %w", err)
+	}
+
+	err = wn.Engine.Reconfig(cfg, wn.router, &dns.Config{}, &tailcfg.Debug{})
+	if err != nil {
+		return xerrors.Errorf("reconfig: %w", err)
+	}
+
+	// Always give the Tailscale engine a copy of our network map.
+	wn.Engine.SetNetworkMap(copyNetMap(wn.netMap))
+	return nil
+}
+
+func copyNetMap(nm *netmap.NetworkMap) *netmap.NetworkMap {
+	nmCopy := *nm
+	return &nmCopy
+}
+
+// Ping sends a discovery ping to the provided peer.
+func (wn *WireguardNetwork) Ping(peer WireguardPeerMessage) *ipnstate.PingResult {
+	ch := make(chan *ipnstate.PingResult)
+	wn.Engine.Ping(peer.IPv6, tailcfg.PingDisco, func(pr *ipnstate.PingResult) {
+		ch <- pr
+	})
+
+	return <-ch
+}
+
+// Listen returns a net.Listener that can be used to accept connections from the
+// wireguard network at the specified address.
```

Comment:
```suggestion
// wireguard network at the specified address. If a listener exists for a given address, all connections will be forwarded to the listener instead of being routed to the host.
```